meteor/server/worker/worker.ts

```diff
@@ -104,6 +102,21 @@ async function getNextJob(queueName: string): Promise<JobSpec | null> {
 	queue.notifyWorker = createManualPromise()
 	return queue.notifyWorker
 }
+/** This is called by each Worker Thread, when it thinks there is a job to execute */
+async function getNextJob(queueName: string): Promise<JobSpec | null> {
+	// Check if there is a job waiting:
+	const queue = getOrCreateQueue(queueName)
+	const job = queue.jobs.shift()
+	if (job) {
+		// If there is a completion handler, register it for execution
+		if (job.completionHandler) runningJobs.set(job.spec.id, job.completionHandler)
+		// Pass the job to the worker
+		return job.spec
+	}
+
+	// No job ready
+	return null
```

Comment:
Is the result of this method ever checked to ensure that it is not null, to e.g prevent NullReferenceException?
Is null checks not something that should be avoided, if possibly?

Returning false, and checking for that should increase the code quality, and potentially improve performance a smidge.
Of cource if the null is NEVER checked, than you can disregard my comment altogether.

Reply:
Returning null or undefined is a bad practice for the exact reason that you stated Krzys. Whenever you call this method you are now forced to make check to see if you actually got the value. 
Everytime you make a method return null or undefined you are setting yourself up for writing extra boilerplate code.

Reply:
But why would false would be any better than null? If it returned false, you would still be forced to check if you got a value.
As @ianshade said, in typescript (unless you disable strict mode which we havent) you have to explicitly declare whether a value can be null. So all calling sites know that it could be null and the compiler will complain if the null isnt handled. 

> Not sure what the desired behaviour of what getNextJob is when there are no jobs, but if it's allowed to wait for a new job when there currently are no jobs, then all you have to do to avoid the null is:

The commit that was cherry-picked across here changed `getNextJob` from blocking to returning immediately, changing it back will reintroduce a bug where jobs can sometimes get lost

Reply:
> It isn't and I didn't claim that. If your return type is Promise<JobSpec> and you return false then you are just downright violating the contract of your method.

You are also violating the contract if the return type is `Promise<JobSpec>` and you return `null` 

> Yes and by doing so you have now forced all callers to write boilerplate code to check for the null - which was my point.

But doing `if (job !== false)` is no different than doing `if (job !== null)`? (Ignoring that these can be written prettier) 

In fact, returning false instead of null or undefined, means you have to do an explicit check, and you can no longer do something like `job?.name`

I would argue that article doesn't apply simply because unless `JobSpec` cannot be null, instead you have to use `JobSpec | null`. Which sounds a lot like their suggestion of `Optional<Item>`. If we define the type, we could even write it like that `type Optional<T> = T | null`

> That was what I thought. Out of curiosity, whereever this is called (I see it dissappears into threadedClass) is it called with await or with .then()?

It is called from https://github.com/nrkno/sofie-core/blob/048cd99795462dca8dcf62805dac320dd49706ed/packages/job-worker/src/workers/parent-base.ts#L262. We are using await because this is run as part of a loop which operates as a workqueue which should only be doing one job at a time. 

Also, `await` in javascript doesn't truly block. It only blocks the current async callstack, execution will switch to somewhere else. So all we are blocking is this job processing while loop.

The basic flow is 
1. `await this.#jobStream.wait()` which 'blocks' until awoken (by a job being queued, or the worker thread status changing, or something else). At this point we havent claimed a job, so if our loop errors or the process exits, the job is still on the queue. 
1. Do some preparation for the job
2. `await this.#jobStream.pop()` fetch the next job if there is one. If there was none, then loop around to 1 (possibly doing some thread status checks first). And because the return type is `Promise<JobSpec | null>`, that helps indicate that it may not return a job. Immediately following that we have a `if  (job) {` which is necessary to make typescript happy

Reply:
@LindvedKrvang IMO that article and the entire sentiment against returning null does simply not apply to TypeScript. Throwing is a bad idea because it does not provide any static checking safety. Returning a “Special Case” object is essentially an equivalent of returning null in TypeScript. With the “Special Case” object approach you also need extra code to check if the returned object was the special case object or not.

Reply:
> Yes and by doing so you have now forced all callers to write boilerplate code to check for the null - which was my point.
If I want to get a result of some method ASAP, but the requested resource might not be there (a queue is empty, or a book with a given title does not exist etc.), I need to be prepared for lack of that resource. I don't want to wait for ages (e.g. until someone publishes a book with that title), so I need some extra code to support null / undefined / exception / rejection etc. in one way or the other. It's just that null is dangerous in Java, hence not recommended, but in TypeScript it isn't.